src/Linux/readInterfaces.c

```diff
@@ -870,10 +870,11 @@ extern "C" {
 	if (ifr.ifr_addr.sa_family == AF_INET) {
 	  struct sockaddr_in *s = (struct sockaddr_in *)&ifr.ifr_addr;
 	  // IP addr is now s->sin_addr
-	  adaptorNIO->ipAddr.type = SFLADDRESSTYPE_IP_V4;
-	  adaptorNIO->ipAddr.address.ip_v4.addr = s->sin_addr.s_addr;
-	  // add to localIP hash too
-	  addLocalIP(newLocalIP, &adaptorNIO->ipAddr, adaptor->deviceName);
+	  SFLAddress ipAddr;
+	  ipAddr.type = SFLADDRESSTYPE_IP_V4;
```

Comment:
`findIpByPriority` may not be necessary.

The current address reading sequence before your change is
1) find ipv4 addr from `ifr.ifr_addr` here
2) `readL3Address` 
3) `readIPv6Addresses`

Leveraging this sequence, If we adopt a `first fit` strategy, it would naturally meet the fix you are targeting `hsflowd will switch the agent to IPv6 address if IPv4 no more available and will back to IPv4 if it will be added to the system`. Specifically:

If we find the current SFLAdaptor is an existing one on line 820, `memset` its `adaptorNIO->ipAddr` field so that `adaptorNIO->ipAddr.type` becomes `SFLADDRESSTYPE_UNDEFINED`. In later process of going through the address reading sequence in this `readInterfaces` function aforementioned, if we find `adaptorNIO->ipAddr.type` is still `SFLADDRESSTYPE_UNDEFINED`, we assign the address to `adaptorNIO->ipAddr` and update `adaptorNIO->ipAddr.type` accordingly.

Due to the address reading sequence, if an ipv4 address is present, it will be selected first. If absent, an ipv6 address will be select, if present, under the prerequisite that `adaptorNIO->ipAddr.type` is still `SFLADDRESSTYPE_UNDEFINED` (i.e., first-fit strategy).

Also, we may need to avoid selecting an ipv6 link local address as it does not have any significance beyond the device scope itself by adding a check `!IN6_IS_ADDR_LINKLOCAL(&v6addr.address.ip_v6)`

Reply:
Thank you for your comments.  I would prefer the address selection not to depend too much upon the discovery implementation.  There are other considerations captured in the priority scheme too,  such as v6 scope (as you mention), RFC1918 address-space preference and possible agent.cidr settings in the config.  Also we want to be free to change how we discover the addresses.  We could decide to rely only on getifaddrs() going forward. Some other platform might require that we learn the set of eligible addresses from a database.  So I think the priority scheme is helpful even if it does seem a bit over-engineered at first glance.